# Imports

In [1]:
import warnings
warnings.filterwarnings('ignore')
from pprint import pprint
import pandas as pd
import numpy as np

%matplotlib inline
import matplotlib.pyplot as plt
import nltk
import spacy
import re
import model
import env
import os

from importlib import reload
from itertools import product
from math import sqrt
from scipy import stats
from scipy.stats import pearsonr, spearmanr
from statsmodels.formula.api import ols
from sklearn.metrics import mean_squared_error, explained_variance_score
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn import tree
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, confusion_matrix, ConfusionMatrixDisplay
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score, precision_recall_curve, auc
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import RFE, SelectKBest, f_regression
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler, StandardScaler, RobustScaler, QuantileTransformer
from sklearn.linear_model import LinearRegression, LassoLars, TweedieRegressor
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer


## Acquire data

In [2]:
#reading the list of prepared pickle file contained in the csv file
df = pd.read_pickle('prepared.pkl')

## Model

After analyzing the Top 200 trending Youtube videos, Exploration phase identified arrays of possible drivers of Top 25 trending Youtube videos. In this section, we will create a machine learning algorithm model that better and accurately predicts Top 25 trending Youtube videos and to use our takeaways and recommendations with an eye towards enabling smaller creators to produce in the mode of the top 25.

Top_25 is our target variable

We will use our training data to train/fit to our model and then tune the model on our validate data.

We will pick our best model on accuracy.

Four supervised machine learning classifications models were created in this project:

 - Decision Tree
 - Randon Forest
 - K-Nearest Neighbor
 - Logistic Regression 

## Split Data
 - Split data into 3 samples of train (60%), validate(20%) and test(20%)
 - Our target variable is Top_25

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2019 entries, 0 to 2018
Data columns (total 37 columns):
 #   Column                   Non-Null Count  Dtype              
---  ------                   --------------  -----              
 0   video_id                 2019 non-null   object             
 1   title                    2019 non-null   object             
 2   publishedAt              2019 non-null   datetime64[ns, UTC]
 3   channelTitle             2019 non-null   object             
 4   categoryId               2019 non-null   object             
 5   trending_date            2019 non-null   datetime64[ns, UTC]
 6   tags                     2019 non-null   object             
 7   view_count               2019 non-null   int64              
 8   likes                    2019 non-null   int64              
 9   comment_count            2019 non-null   int64              
 10  thumbnail_link           2019 non-null   object             
 11  comments_disabled        2019 

In [4]:
dummy_df = pd.get_dummies(df[['categoryId','region']], drop_first=True)

In [5]:
df = pd.concat([df, dummy_df], axis=1)

In [6]:
df_copy = df.copy()

In [7]:
df = df.drop(columns=['categoryId','video_id','title','publishedAt','region','trending_date','tags','channelTitle',
                'thumbnail_link','comments_disabled','ratings_disabled','ratings_disabled',
                'ratings_disabled','description','captions','channel_age', 'rank',
                'word_bank','cleaned_tags','cleaned_desc','title_in_description','title_in_tags'],\
                axis=1)

df

view_count   likes  comment_count  duration  top_25  subscribers  \
0        1098919   19090            861      1281       1      2210000   
1        2217807  182434           7282      4327       1      7180000   
2        2258144   44366            938     24001       1     19200000   
3        3262953  312903          14437      1862       1     15000000   
4         313545    4899            596       965       1       795000   
...          ...     ...            ...       ...     ...          ...   
2014     1902762   78690           1712      1099       0     12400000   
2015     1976426   48816           3060      1869       0      4630000   
2016     1529830  114260          11625       166       0      7720000   
2017     1835677   12011           1015      1191       0       321000   
2018      430118   37109           1313       482       0      2570000   

      video_count         age  engagement  sponsored  ...  region_CA  \
0            3168   82.554444    0.020506          0  ...          0   
1             311   29.206667    0.095392          0  ...          1   
2             603   15.998889    0.021309          0  ...          0   
3             408   27.088611    0.113594          0  ...          0   
4            2211    3.805000    0.023228          0  ...          0   
...           ...         ...         ...        ...  ...        ...   
2014         1270  270.941944    0.044955          0  ...          0   
2015         1829  228.996944    0.030892          0  ...          0   
2016          206  285.015833    0.105084          0  ...          0   
2017           81  566.429167    0.008755          0  ...          0   
2018         2608  101.242222    0.098487          0  ...          0   

      region_DE  region_FR  region_GB  region_IND  region_JP  region_KR  \
0             0          0          0           0          0          1   
1             0          0          0           0          0          0   
2             1          0          0           0          0          0   
3             0          0          0           0          0          0   
4             0          0          0           0          1          0   
...         ...        ...        ...         ...        ...        ...   
2014          0          0          0           0          0          0   
2015          0          0          0           0          1          0   
2016          0          0          0           0          0          0   
2017          0          0          0           0          0          1   
2018          0          1          0           0          0          0   

      region_MX  region_RU  region_US  
0             0          0          0  
1             0          0          0  
2             0          0          0  
3             0          1          0  
4             0          0          0  
...         ...        ...        ...  
2014          1          0          0  
2015          0          0          0  
2016          0          0          0  
2017          0          0          0  
2018          0          0          0  

[2019 rows x 41 columns]

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2019 entries, 0 to 2018
Data columns (total 41 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   view_count                      2019 non-null   int64  
 1   likes                           2019 non-null   int64  
 2   comment_count                   2019 non-null   int64  
 3   duration                        2019 non-null   int64  
 4   top_25                          2019 non-null   int64  
 5   subscribers                     2019 non-null   int64  
 6   video_count                     2019 non-null   int64  
 7   age                             2019 non-null   float64
 8   engagement                      2019 non-null   float64
 9   sponsored                       2019 non-null   int64  
 10  num_of_tags                     2019 non-null   int64  
 11  pct_tags_in_description         2019 non-null   float64
 12  title_lengths                   20

In [9]:
train, validate, test = model.my_train_test_split(df, 'top_25')

In [10]:
train.shape,validate.shape,test.shape

((1211, 41), (404, 41), (404, 41))

## Scale Data

In [11]:
#dummy_df = pd.get_dummies(df[['categoryId','region']],drop_first=True)
#dummy_df.head()

In [12]:
#pd.get_dummies(df, prefix=['col1', 'col2'])

In [13]:
#pd.get_dummies(df, prefix=['categoryId', 'region'])

In [14]:
from sklearn.preprocessing import MinMaxScaler

In [15]:
#selected features to scale into train, validate & test
features_to_scale = ['view_count', 'likes', 'comment_count', 'duration',\
            'num_of_tags', 'pct_tags_in_description',\
            'title_lengths','desc_lengths','tags_length',\
            'content_rate', 'views_per_sub',\
            'subscribers','video_count','age','engagement','sponsored']

In [16]:
train_scaled, validate_scaled, test_scaled = model.scale_data(train, validate, test, features_to_scale)

## Feature Selection
- Selected 16 best
- Uses statistical tests to determine each feature's usefulness in predicting the target variable.
- Ranks the features based of highest correlation to engagement.

In [17]:
#X will be features
#y will be our target variable
#these features have high correlation to top_25 videos
X_train, y_train, X_validate, y_validate, X_test, y_test = model.getting_(train_scaled,validate_scaled,test_scaled)

## Baseline Prediction and Accuracy
- Baseline prediction is a benchmark. It predicts the most prevelant class in the train data. We compare our model and want it to be better than the baseline prediction.

In [18]:
# look at values of target variable top_25
# baseline prediction: the most prevalent class in training dataset(the mode)
y_train.value_counts()

0    987
1    224
Name: top_25, dtype: int64

In [19]:
#Formulate baseline accuracy
baseline_accuracy = (y_train == 0).mean()
print('Top_25 videos baseline accuracy is:', baseline_accuracy)

Top_25 videos baseline accuracy is: 0.815028901734104


## Model on Train

## Decision Tree Classifier

In [20]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1211 entries, 231 to 544
Data columns (total 56 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   view_count                      1211 non-null   int64  
 1   likes                           1211 non-null   int64  
 2   comment_count                   1211 non-null   int64  
 3   duration                        1211 non-null   int64  
 4   subscribers                     1211 non-null   int64  
 5   video_count                     1211 non-null   int64  
 6   age                             1211 non-null   float64
 7   engagement                      1211 non-null   float64
 8   sponsored                       1211 non-null   int64  
 9   num_of_tags                     1211 non-null   int64  
 10  pct_tags_in_description         1211 non-null   float64
 11  title_lengths                   1211 non-null   int64  
 12  desc_lengths                    1

In [21]:
dtc_scores = model.run_decision_tree_models(X_train, y_train, X_validate, y_validate)

In [22]:
dtc_scores

i  accuracy_train  accuracy_validate  difference
0    1        0.815029           0.814356    0.000672
1    2        0.853014           0.831683    0.021331
2    3        0.867878           0.841584    0.026294
3    4        0.889348           0.883663    0.005684
4    5        0.922378           0.876238    0.046141
5    6        0.943848           0.876238    0.067610
6    7        0.958712           0.873762    0.084949
7    8        0.971098           0.878713    0.092385
8    9        0.986788           0.861386    0.125402
9   10        0.995045           0.871287    0.123758
10  11        0.997523           0.856436    0.141087

In [23]:
#model with maximun sample leaf 6 @ 85% accuracy on validate is the best

## Random Forest

In [24]:
from importlib import reload

In [25]:
reload(model)

<module 'model' from '/Users/sinao/codeup-data-science/top_200/model.py'>

In [26]:
rf_scores = model.run_random_forest_models(X_train, y_train, X_validate, y_validate)

In [27]:
rf_scores

max_depth  accuracy_train  accuracy_validate  difference
0           5        0.900083           0.853960    0.046122
1           6        0.914946           0.861386    0.053560
2           7        0.935590           0.863861    0.071729
3           8        0.951280           0.873762    0.077518
4           9        0.966969           0.871287    0.095682
5          10        0.975227           0.876238    0.098989
6          11        0.984310           0.876238    0.108073
7          12        0.988439           0.876238    0.112202
8          13        0.988439           0.873762    0.114677
9          14        0.988439           0.883663    0.104776
10         15        0.991742           0.886139    0.105604
11         16        0.990091           0.881188    0.108903
12         17        0.992568           0.881188    0.111380
13         18        0.990091           0.881188    0.108903
14         19        0.990091           0.881188    0.108903

In [28]:
#model with maximun sample leaf 10 @ 84% accuracy on validate is the best

## KNeighborsClassifier

In [29]:
reload(model)

<module 'model' from '/Users/sinao/codeup-data-science/top_200/model.py'>

In [30]:
knn_scores = model.run_kneighbors_models(X_train, y_train, X_validate, y_validate)

In [31]:
knn_scores

k  accuracy_train  accuracy_validate  difference
0  1        1.000000           0.690594    0.309406
1  2        0.853840           0.811881    0.041959
2  3        0.853014           0.782178    0.070836
3  4        0.830718           0.809406    0.021312
4  5        0.831544           0.806931    0.024613
5  6        0.826590           0.806931    0.019659
6  7        0.824112           0.804455    0.019657
7  8        0.820809           0.809406    0.011403
8  9        0.819158           0.801980    0.017178

In [32]:
#KNN model with minimun sample leaf 9 @ 80% accuracy on validate is the best

## Logistic Regression

In [33]:
reload(model)

<module 'model' from '/Users/sinao/codeup-data-science/top_200/model.py'>

In [34]:
lr_scores = model.run_logistic_reg_models(X_train, y_train, X_validate, y_validate)

In [35]:
lr_scores

{'accuracy_train': 0.8142031379025598,
 'accuracy_validate': 0.8217821782178217,
 'difference': -0.007579040315261909}

In [36]:
#Logistic Regression has 81% accuracy on validate

## Test Model

In [40]:
def run__on_test(X_train, y_train, X_test, y_test):
    #create, fit, use, model information to model_features dfram
    model = DecisionTreeClassifier(max_depth=5, random_state=123)
    #features to be used

    scaled_features = ['age','duration','num_of_tags','engagement','sponsored', 'title_in_description', 
        'title_in_tags','pct_tags_in_description', 'title_lengths', 'desc_lengths','tags_length',
                     'channel_age','subscribers','video_count','content_rate','views_per_sub']
    #fit model
    model.fit(X_train, y_train)
    #score model to add to model description dataframe
    score = model.score(X_test, y_test).round(3)
    
    return score

In [41]:
run__on_test(X_train, y_train, X_test, y_test)

0.859

In [42]:
model.run__on_test(X_train, y_train, X_test, y_test)

0.854

## Test Takeaways

- Decision Tree Classifier model performed best on validate data with 94% accuracy, so we will use it on test data

## Modeling Takeaways

- All models had roughly almost the same accuracy for train and validate sets.
- Logistic Regression model performed the worst on out-of-sample data.
- The best performing model is Decision Tree Classifier.
- However, from the above model analysis we can see that the highest accuracy is achieved by Decision Tree. It performs better than baseline by about 7%.

- While this is an improvement there is still room for improvement in future iterations